### PR TITLE
potential bugfix for disjoint smartgraph queries that refer to collections on other servers

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Potential bugfix/workaround for issues with disjoint smart graph queries that have a "WITH" declaration and that can refer to other collections/shards, on different servers.

Given a query such as
```
WITH a, c
FOR v, e IN 1 .. 1 ANY "a/11:82238118" b 
  FOR doc in c
    FILTER doc.color == e.color
    RETURN doc
```
and collections with the following setup:

* `a` is a disjoint smart vertex collection, with 3 shards, on 3 different servers
* `b` is a disjoint smart edge collection, with 3 shards, on 3 different servers
* `c` is an unrelated collection, with just 1 shard, on only one server

The problem here is that `a` and `b` are distributed in the same way, but `c` is not. Especially, on some servers that there will be no shards for `c`.

Now, when the query snippets are created, all extra vertex and edge collections will be enumerated and their shards will be added. Here, this is `c`. But on some servers there are no shards for `c`, so there will be no entry in the collection->shard map for it.

This PR inserts a placeholder for such collections which have no shard attached, and on lookup just ignores them. This seems to work in my quick tests, and prevents the coordinator from running into undefined behavior.
The PR also returns a proper error in case no shard is mapped for a collection, so that it is clear what happened and the coordinator can live on.

It is unclear to me if this is a proper way to address the problem, as my knowledge about the subject is only 5 minutes of code-reading and poking with [the one true debugger](https://news.ycombinator.com/item?id=25858969).

If this fix is considered ok, we will need to add tests for this particular case, and port the fix to devel, too.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13954/

### CODE REVIEW

**Comment by mchacki** [comment]
Jenkins:
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13966/

Enterprise companion:
https://github.com/arangodb/enterprise/pull/648

**Comment by mchacki** [comment]
Jenkins after JSLint fixes:
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13967/

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,30 @@
...
+  the AQL setup to create erroneous state.
+  This is now fixed and the above query works.
+  This query:
+
+  WITH vertices, unrelated
+  FOR out IN 1 OUTBOUND "v/1:1" GRAPH "graph"
+    FOR u IN unrelated
+    RETURN [out, u]
+
+  was not effected by this bug.
```
Comment (goedderz) [inline CHANGELOG:26]:
```suggestion
  was not affected by this bug.
```

File: arangod/Aql/GraphNode.cpp [hunk 2]
```diff
@@ -508,6 +508,9 @@ std::string const& GraphNode::collectionToShardName(std::string const& collName)
   }
 
   auto found = _collectionToShard.find(collName);
+  if (found == _collectionToShard.end()) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "unable to find shard '" + collName + "' in query shard map");
```
Comment (goedderz) [inline arangod/Aql/GraphNode.cpp:512]:
```suggestion
    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL_AQL, "unable to find shard '" + collName + "' in query shard map");
```

File: arangod/Aql/GraphNode.cpp [hunk 3]
```diff
@@ -508,6 +508,9 @@ std::string const& GraphNode::collectionToShardName(std::string const& collName)
...
+  }
   TRI_ASSERT(found != _collectionToShard.cend());
```
Comment (goedderz) [inline arangod/Aql/GraphNode.cpp:514]:
```suggestion
```

File: arangod/Aql/QuerySnippet.cpp [hunk 4]
```diff
@@ -668,14 +668,24 @@ auto QuerySnippet::prepareFirstBranch(
...
+            } else {
+              // the target server does not have anything to do with the particular
+              // collection (e.g. because the collection's shards are all on other
+              // servers), but it may be asked for this collection, because vertex
+              // collections are registered _globally_ with the TraversalNode and
+              // not on a per-target server basis.
+              // so in order to serve later lookups for this collection, we insert
+              // an empty string into the collection->shard map.
+              // on lookup, we will react to this.
+              localGraphNode->addCollectionToShard(aqlCollection->name(), "");
```
Comment (goedderz) [inline arangod/Aql/QuerySnippet.cpp:680]: Changing the `_collectionToShard` map to contain `optional<string>`s would make it a little more explicit, but its not necessary.

**Review by goedderz — CHANGES_REQUESTED** [review]
I think we should add a test for this.

**Review by goedderz — APPROVED** [review]

File: arangod/Aql/QuerySnippet.cpp (hunk 4 above)
Comment (mchacki) [reply arangod/Aql/QuerySnippet.cpp:680]: I think this is just left here as a security net, this should never show up anymore.

**Comment by mchacki** [comment]
Jenkins after merge from Review:
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13972/

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/7429/

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/7444/

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/7449/

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/7453/

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/7454/

**Comment by KVS85** [comment]
Tests blue except unrelated known failure.

**Pull request merged**